cmd/diff/main.go

```diff
@@ -0,0 +1,290 @@
+		toNode := toManifest.GetFileNode(path)
+		if toNode == nil {
+			diff.removed(path, fromNode)
+			continue
+		}
+		if bufcas.DigestEqual(fromNode.Digest(), toNode.Digest()) {
+			continue // no changes
+		}
+		if err := diff.changedDigest(ctx, bucket, path, fromNode, toNode); err != nil {
+			return fmt.Errorf(
+				"changed digest from %s to %s: %w",
+				fromNode.String(),
+				toNode.String(),
+				err,
+			)
+		}
+	}
+	// added
+	for _, toNode := range toManifest.FileNodes() {
+		if fromManifest.GetFileNode(toNode.Path()) == nil {
+			diff.added(toNode.Path(), toNode)
+		}
+	}
+	_, _ = fmt.Fprint(os.Stdout, diff.String())
+	return nil
+}
+
+func readManifest(ctx context.Context, bucket storage.ReadBucket, manifestPath string) (bufcas.Manifest, error) {
+	data, err := storage.ReadPath(ctx, bucket, manifestPath)
+	if err != nil {
+		return nil, fmt.Errorf("read path: %w", err)
+	}
+	m, err := bufcas.ParseManifest(string(data))
+	if err != nil {
+		return nil, fmt.Errorf("parse manifest: %w", err)
+	}
+	return m, nil
+}
+
+type fileNodeDiff struct {
+	path string
+	from bufcas.Digest
+	to   bufcas.Digest
+	diff string
+}
+
+type manifestDiff struct {
+	addedPaths         map[string]bufcas.FileNode
+	removedPaths       map[string]bufcas.FileNode
+	changedDigestPaths map[string]fileNodeDiff
+}
+
+func newManifestDiff() *manifestDiff {
+	return &manifestDiff{
+		addedPaths:         make(map[string]bufcas.FileNode),
+		removedPaths:       make(map[string]bufcas.FileNode),
+		changedDigestPaths: make(map[string]fileNodeDiff),
+	}
+}
+
+func (d *manifestDiff) added(path string, node bufcas.FileNode) {
+	d.addedPaths[path] = node
+}
+
+func (d *manifestDiff) removed(path string, node bufcas.FileNode) {
+	d.removedPaths[path] = node
+}
+
+func (d *manifestDiff) changedDigest(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+	path string,
+	fromNode bufcas.FileNode,
+	toNode bufcas.FileNode,
+) error {
+	if fromNode.Path() != toNode.Path() {
+		return errors.New("from path and to path are different")
+	}
+	fromData, err := storage.ReadPath(ctx, bucket, fileNodePath(fromNode))
+	if err != nil {
+		return fmt.Errorf("read from changed path: %w", err)
+	}
+	toData, err := storage.ReadPath(ctx, bucket, fileNodePath(toNode))
+	if err != nil {
+		return fmt.Errorf("read to changed path: %w", err)
+	}
+	diffData, err := diff.Diff(
+		ctx,
+		fromData,
+		toData,
+		fromNode.Digest().String(),
+		toNode.Digest().String(),
+		diff.DiffWithSuppressCommands(),
+		diff.DiffWithSuppressTimestamps(),
+	)
+	if err != nil {
+		return fmt.Errorf("diff: %w", err)
+	}
+	d.changedDigestPaths[path] = fileNodeDiff{
+		path: fromNode.Path(),
+		from: fromNode.Digest(),
+		to:   toNode.Digest(),
+		diff: string(diffData),
+	}
+	return nil
+}
+
+func (d *manifestDiff) String() string {
+	var sb strings.Builder //nolint:varnamelen // sb is a common builder varname
+	if len(d.removedPaths) > 0 {
+		sb.WriteString("# Removed:\n\n")
+		sb.WriteString("```diff\n")
```

Comment:
It would be nice if there was a `--format={text,markdown}` option here, with the default being a unified diff in text format. Supporting markdown is a nice addition for the purposes of the modules release notes.

Reply:
Fair. Since the diff is more like a "directory diff", how would a unified diff in text format would look like? Any suggestions?

Reply:
Basically the same just without the Markdown headings and code blocks.